base/compiler/inferencestate.jl

```diff
@@ -1116,3 +1121,74 @@ function get_max_methods_for_module(mod::Module)
     max_methods < 0 && return nothing
     return max_methods
 end
+
+"""
+    Future{T}
+
+Delayed return value for a value of type `T`, similar to RefValue{T}, but
+explicitly represents completed as a `Bool` rather than as `isdefined`.
+Set once with `f[] = v` and accessed with `f[]` afterwards.
+
+Can also be constructed with the `completed` flag value and a closure to
+produce `x`, as well as the additional arguments to avoid always capturing the
+same couple of values.
+"""
+mutable struct Future{T}
+    x::T
+    completed::Bool
+    Future{T}() where {T} = (f = new{T}(); f.completed = false; f)
+    Future{T}(x) where {T} = new{T}(x, true)
+    Future(x::T) where {T} = new{T}(x, true)
+end
+getindex(f::Future) = (@assert f.completed; f.x)
+setindex!(f::Future, v) = (@assert !f.completed; f.completed = true; f.x = v; f)
+convert(::Type{Future{T}}, x) where {T} = Future{T}(x) # support return type conversion
+convert(::Type{Future{T}}, x::Future) where {T} = x::Future{T}
```

Comment:
Before merging this PR, I'd like to make sure the code works without this `convert`.

Reply:
The `convert` is very useful convenience syntax, since it means return-types work